tests/check_public_api.py

```diff
@@ -0,0 +1,131 @@
+import tempfile
+import os
+import re
+
+
+def get_public_symbols_from_shared_object(prefix):
+    """
+    Find public (exported) symbols in a shared library (*.so) by calling
+    the 'nm' command.
+    :param prefix: CMAKE_INSTALL_PREFIX
+    :return: a set of public symbols
+    """
+    so_path = os.path.join(prefix, "lib64", "libopenscap.so")
+    nm_command = ["nm", "--demangle", "--dynamic", "--defined-only", so_path]
+    nm_output_bytes = subprocess.check_output(nm_command)
+    nm_output = nm_output_bytes.decode(encoding="utf-8")
+    symbols = {line.split()[2] for line in nm_output.splitlines()}
+    return symbols
+
+
+def get_public_headers(prefix):
+    """
+    It is useful to check if all the public headers are actually installed under $PREFIX/include/openscap.
+    :param prefix: CMAKE_INSTALL_PREFIX
+    :return: a set of public headers filepaths
+    """
+    public_headers = set()
+    include_path = os.path.join(prefix, "include", "openscap")
+    for root, dirs, files in os.walk(include_path):
+        for f in files:
+            full_path = os.path.join(root, f)
+            public_headers.add(full_path)
+    return public_headers
+
+
+def get_public_symbols_from_header(header_path):
+    """
+    Parses a header file to find symbols marked by OSCAP_API macro.
+    :param header_path: path to the header file
+    :return: a list of public symbols in the header file
+    """
+    symbols = set()
+    prototype_regex = re.compile(r"^[ \t]*OSCAP_API[^;]*;", re.MULTILINE)
+    with open(header_path, "r") as header_file:
+        content = header_file.read()
+    prototypes = prototype_regex.findall(content)
+    function_name_regex = re.compile(r"([a-zA-Z0-9_]+)\s*\(.*\)")
+    for p in prototypes:
+        p = p.replace("\n", " ")
+        f_match = function_name_regex.search(p)
+        f_name = f_match.group(1) if f_match else "unknown"
+        symbols.add(f_name)
+    return symbols
+
+
+def get_public_symbols_from_headers(headers):
+    """
+    Find public (exported) symbols in header files by finding symbols
+    marked as OSCAP_API.
+    :param headers: a list of public headers file paths
+    :return: a set of public symbols in public headers
+    """
+    symbols = set()
+    for h in headers:
+        symbols.update(get_public_symbols_from_header(h))
+    return symbols
+
+
+def analyze_project_artifacts(src_dir):
+    """
+    Analyze the project and find all the symbols declared as public in
+    public header files and all the symbols actually exported in built
+    library. This is done on a real build and a real installation.
+    The artifacts are installed in a temporary directory which is
+    automatically removed when this function returns.
+    :param src_dir: Path to project sources
+    :return: tuple (header_symbols, so_symbols)
+    """
+    cwd = os.getcwd()
+    with tempfile.TemporaryDirectory() as prefix:
+        build_dir = os.path.join(src_dir, "build")
+        os.chdir(build_dir)
```

Comment:
I think it would be good to add a sanity check here for when `build` is non-empty or perhaps when the build was configured with `ninja` instead of `make`.

That's a non-trivial gotcha that some people will run into that is easily detected by this script (`build.ninja`) and causes an opaque error:

```
-- OpenSCAP 1.3.0_alpha3
-- (see /home/cipherboy/GitHub/jan-cerny/openscap/docs/developer/developer.adoc for build instructions)
--  
-- Checking for module 'acl'
--   Package 'acl', required by 'virtual:world', not found
-- Found ACL 
-- Found BLKID 
-- Found GLib 
-- Found GObject 
-- Found GCONF 
-- Checking for module 'opendbx'
--   Package 'opendbx', required by 'virtual:world', not found
-- Found POPT 
-- Checking for module 'libprocps'
--   Package 'libprocps', required by 'virtual:world', not found
-- Found RPM 4.14.1
-- Found SELINUX 
--  
-- CMake:
-- generator: Ninja
-- source directory: /home/cipherboy/GitHub/jan-cerny/openscap
-- build directory: /home/cipherboy/GitHub/jan-cerny/openscap/build
--  
-- Core features:
-- SCE: OFF
--  
-- OVAL:
-- base probe support: ON
-- SEAP msgid bit-size: 32
-- independent probes: ON
-- unix probes: ON
-- linux probes: ON
-- solaris probes: OFF
--  
-- Language bindings:
-- python3 bindings: FALSE
-- perl bindings: FALSE
--  
-- Utilities:
-- oscap: ON
-- oscap-docker: ON
-- scap-as-rpm: ON
-- oscap-ssh: ON
-- oscap-vm: ON
-- oscap-chroot: ON
--  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/cipherboy/GitHub/jan-cerny/openscap/build
make: *** No targets specified and no makefile found.  Stop.
make: Nothing to be done for 'docs'.
make: *** No rule to make target 'install'.  Stop.
nm: '/tmp/tmp4udeo22_/lib64/libopenscap.so': No such file
Traceback (most recent call last):
  File "../tests/check_public_api.py", line 130, in <module>
    main()
  File "../tests/check_public_api.py", line 110, in main
    header_symbols, so_symbols = analyze_project_artifacts(src_dir)
  File "../tests/check_public_api.py", line 102, in analyze_project_artifacts
    so_symbols = get_public_symbols_from_shared_object(prefix)
  File "../tests/check_public_api.py", line 18, in get_public_symbols_from_shared_object
    nm_output_bytes = subprocess.check_output(nm_command)
  File "/usr/lib64/python3.6/subprocess.py", line 336, in check_output
    **kwargs).stdout
  File "/usr/lib64/python3.6/subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['nm', '--demangle', '--dynamic', '--defined-only', '/tmp/tmp4udeo22_/lib64/libopenscap.so']' returned non-zero exit status 1.
```

In addition, you might want to check the return codes of the `make` commands so they don't all get called if they are failing.